graphbolt/src/cuda/index_select_impl.cu

```diff
@@ -48,6 +53,294 @@ std::pair<torch::Tensor, torch::Tensor> Sort(
+          using indices_t = decltype(indices_type_variable);
+
+          // Actual and modified number of edges.
+          indptr_t hop_size, hop_size_aligned;
+          auto sub_indptr_aligned =
+              allocator.AllocateStorage<indptr_t>(num_rows + 1);
+          {
+            thrust::counting_iterator<int64_t> iota(0);
+            // Returns the actual and modified_indegree as a pair, the latter
+            // overestimates the actual indegree for alignment purposes.
+            auto modified_in_deg = thrust::make_transform_iterator(
+                iota, AlignmentFunc<indptr_t, indices_t>{
+                          in_deg.get(), perm, num_rows});
+            auto sub_indptr_pair = thrust::make_zip_iterator(
+                sub_indptr.data_ptr<indptr_t>(), sub_indptr_aligned.get());
+            thrust::tuple<indptr_t, indptr_t> zero_value{};
+            // Compute the prefix sum over actual and modified indegrees.
+            size_t workspace_size = 0;
+            CUDA_CALL(cub::DeviceScan::ExclusiveScan(
+                nullptr, workspace_size, modified_in_deg, sub_indptr_pair,
+                PairSum{}, zero_value, num_rows + 1, stream));
+            auto temp = allocator.AllocateStorage<char>(workspace_size);
+            CUDA_CALL(cub::DeviceScan::ExclusiveScan(
+                temp.get(), workspace_size, modified_in_deg, sub_indptr_pair,
+                PairSum{}, zero_value, num_rows + 1, stream));
+          }
+          // Copy the modified number of edges.
+          CUDA_CALL(cudaMemcpyAsync(
+              &hop_size_aligned, sub_indptr_aligned.get() + num_rows,
+              sizeof(hop_size_aligned), cudaMemcpyDeviceToHost, stream));
+          // Copy the actual total number of edges.
+          CUDA_CALL(cudaMemcpyAsync(
+              &hop_size, sub_indptr.data_ptr<indptr_t>() + num_rows,
+              sizeof(hop_size), cudaMemcpyDeviceToHost, stream));
+          // synchronizes here, we can read hop_size and hop_size_aligned
+          CUDA_CALL(cudaStreamSynchronize(stream));
+          // Allocate output array with actual number of edges.
+          sub_indices = torch::empty(
+              hop_size, nodes.options().dtype(indices.scalar_type()));
+          const dim3 block(BLOCK_SIZE);
+          const dim3 grid((hop_size_aligned + BLOCK_SIZE - 1) / BLOCK_SIZE);
+          // Perform the actual copying, of the indices array into sub_indices
+          // in an aligned manner.
+          CUDA_KERNEL_CALL(
+              _CSRRowWiseOneHopExtractorAlignedKernel, grid, block, 0, stream,
+              hop_size_aligned, num_rows, sliced_indptr.get(),
+              sub_indptr.data_ptr<indptr_t>(), sub_indptr_aligned.get(),
+              reinterpret_cast<indices_t*>(indices.data_ptr()),
+              reinterpret_cast<indices_t*>(sub_indices.data_ptr()), perm);
+        };
+        // Dispatch here depending on the sizeof the indices tensor dtype.
+        switch (indices.element_size()) {
```

Comment:
Here, I am doing the dispatch with respect to the sizeof the dtype, similar to how it was done in the UVAIndexSelectImpl. There is no torch dispatch macro for what I would like to do here.

Can you clarify what exactly I should change here?

Reply:
Yeah. Lambda makes the code extremely complicated within a function. So I am suggesting to use global functions as many as possible so the main function is cleaner. Another issue is the data type switching, which can be implemented by template For example, 

```
switch (indices.element_size()) {
    case 1:
        using DType = uint8_t;
        CopyIndices<DType>(...);
}
```

Reply:
I did something different, I added a dispatch macro so that we don't have to use named lambdas, the dispatch is just like torch dispatch functions now. The code should be more readable as a result.

Reply:
Please let me know what you think.